07_errors/jimbotech/sync_mapstore.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"math/rand"
+	"sync"
+)
+
+// SyncMapStore stores puppies threadsafe.
+type SyncMapStore struct {
+	mux sync.Mutex
+	sync.Map
+}
+
+// length is not concorrency safe. As the go doc says:
+// Range does not necessarily correspond to any consistent snapshot of the
+// Map's contents: no key will be visited more than once, but if the value
+// for any key is stored or deleted concurrently, Range may reflect any
+// mapping for that key from any point during the Range call.
+//
+func (s *SyncMapStore) length() int {
+	var length int
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.Range(func(key interface{}, value interface{}) bool {
+		length++
+		return true
+	})
+	return length
+}
+
+// CreatePuppy threadsafe adding a puppy to storage
+// but will modify the member ID.
+func (s *SyncMapStore) CreatePuppy(p *Puppy) (int32, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	p.ID = rand.Int31()
+	sp := *p
+	s.Store(p.ID, &sp)
+	return p.ID, nil
+}
+
+// ReadPuppy threadsafe retrieval of your puppy.
+func (s *SyncMapStore) ReadPuppy(id int32) (*Puppy, error) {
+	if id < 0 {
+		return nil, ErrValueBelowZero
+	}
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	val, found := s.Load(id)
+	if !found {
+		return nil, ErrIDNotFound
+	}
+	retPup := *val.(*Puppy)
+	return &retPup, nil
+}
+
+// UpdatePuppy threadsafe update your puppy store.
+func (s *SyncMapStore) UpdatePuppy(id int32, puppy *Puppy) error {
+	_, err := s.ReadPuppy(id)
+	if err != nil {
+		return err
+	}
+	s.mux.Lock()
```

Comment:
since `sync.Map` is already concurrency safe for individual operations, the only purpose of having a mutex as well is to provide concurrency safety over multiple operations on the map. Yet here you do not take the lock until after you do `ReadPuppy`. That takes and releases the lock, meaning the lock is not held over the duration of all the operations on the `sync.Map`.

The only way this can work then is to not use `ReadPuppy` in this method - you will need to take the lock and directly retrieve from the `sync.Map`

Reply:
Not sure what motivated me to do the code this way. Too long ago. The ReadPuppy is called in order to produce the required errors. If I do what you suggest and directly retrieve from the the sync.Map I would have to replicate a lot of code. I guess I put the lock where I did as I wanted to hold it for the shortest possible time, but as you pointed out, another thread can get in between Read and the change. I will move the lock to the top of the function and separate out the error generating retrieval function into unsafeRetrievePuppy.